EIPS/eip-4762.md

```diff
@@ -35,155 +35,266 @@ def get_storage_slot_tree_keys(storage_key: int) -> [int, int]:
-     * Example 1: for a `CODECOPY` with start position 100, read size 50, `code_size = 200`, `x = 100` and `y = 149`
-     * Example 2: for a `CODECOPY` with start position 600, read size 0, no chunks are accessed
-     * Example 3: for a `CODECOPY` with start position 1500, read size 2000, `code_size = 3100`, `x = 1500` and `y = 3099`
- * `CODESIZE`, `EXTCODESIZE` and `EXTCODEHASH` do NOT access any chunks.
-    When a contract is created, access chunks `0 ... (len(code)+30)//31`
+##### Access events for delegation
 
-### Write Events
+When `contract_address` is that of a delegating EoA, process the following events:
 
-We define **write events** as follows. Note that when a write takes place, an access event also takes place (so the definition below should be a subset of the definition of access events). A write event is of the form `(address, sub_key, leaf_key)`, determining what data is being written to.
-
-#### Write events for account headers
-
-When a nonzero-balance-sending `CALL`, `CALLCODE` or `SELFDESTRUCT` with a given sender and recipient takes place, process these write events:
+ - For an `EXTCODEHASH`:
 
 ```
-(caller, 0, BASIC_DATA_LEAF_KEY)
-(callee, 0, BASIC_DATA_LEAF_KEY)
+(delegated_address, 0, CODEHASH_LEAF_KEY)
 ```
 
-if no account exists at `callee_address`, also process:
+ - For an `EXTCODESIZE`, or any other `*CALL`, `CALLCODE`:
 
 ```
-(callee, 0, CODEHASH_LEAF_KEY)
+(delegated_address, 0, BASIC_DATA_LEAF_KEY)
 ```
 
-When a contract creation is initialized, process these write events:
+#### Access events for code
 
+```python3
+def interpreter_loop(code: bytes, contract_address: bytes20):
+  pc = 0
+
+  # ...
+
+  while True:
+    if pc < len(code):
+      emit_code_access_events(pc, code, contract_address)
+      # ... rest of the interpreter loop
+
+# Executed when decoding an instruction when running the interpreter loop
+def emit_code_access_events(pc: int, code: bytes, contract_address: Address):
+  opcode = OP_STOP # pad chunk with 0s
+  # if the code walks beyond the code size, it is not considered to be accessed.
+  # XXX I don't think that this is true on mainnet, even before the atomic stuff!
+  if pc >= len(code):
+    return
+
+  opcode = code[pc]
+  chunk = pc // CHUNK_SIZE
+
+  add_read_access_event(contract_addresss, (chunk + CODE_OFFSET) // 256, (chunk + CODE_OFFSET) % 256)
+  charge_gas() # Note that the witness gas for instruction access is charged immediately,
+               # independently of access costs for its immediates.
+
+  # add access events for immediates, if the instruction has them
+  match opcode:
+    case OP_JUMP:
+      dest = stack.peek(0)
+      # The destination of a jump is not considered to be accessed if it is beyond the code
+      if dest < len(code):
+        # The destination of a `JUMP` is considered to be accessed, even if the destination is not a jumpdest or is inside pushdata
+        add_read_access_event(dest)
```

Comment:
To fit the signature of the method defined at the top, this is missing two params: subkey, and leaf_key